### PR TITLE
Ftr: add collection support according to java implementation

### PR DESCRIPTION
**Pull request by YGrylls**
**What this PR does**:
Add support for java collections such as HashSet
Collections are encoded through Serializer and decoded as list with regard of its type (collection types are encoded as the type of FixedTypedList in java impl of hessian). 
The collection struct implements are open for users, see what java_collection_test.go does.
Issue #122 

**Special notes for your reviewer**:
To reach the feature, code of method "readTypedList" in list.go also changes a little.
Related test codes of java side have also been added, see the last methods in TestCustomDecode.java & TestCustomReply.java

**Does this PR introduce a user-facing change?**:

**Users need to implements collection structs and set them into go-hessian2, examples can be found in java_collection_test.go**

### CODE REVIEW

**Comment by AlexStocks** [comment]
pls update readme about this feature.

**Comment by AlexStocks** [comment]
@hxmhlt @fangyincheng pls own this pr.

File: java_collection_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+func (j *JavaHashSet) Set(v []interface{}) {
+	j.value = v
+}
+
+func (j *JavaHashSet) JavaClassName() string {
+	return "java.util.HashSet"
+}
+
+func init() {
```
Comment (fangyincheng) [inline java_collection_test.go]: top `init()`?

Comment (YGrylls) [reply java_collection_test.go]: fixed

File: java_collection_test.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"testing"
+)
+
+func init() {
+	SetCollectionSerialize(&JavaHashSet{})
+}
+
+type JavaHashSet struct {
```
Comment (aliiohs) [inline java_collection_test.go]: 在gost中我们已经实现了正在的hashset，这个是否可以复用？

Comment (YGrylls) [reply java_collection_test.go]: 这边设计设想是依赖接口，集合的具体实现开放给用户或外部来实现；这里只是一个实现了接口的简单例子

File: java_collection.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package hessian
+
+import (
+	perrors "github.com/pkg/errors"
```
Comment (AlexStocks) [inline java_collection.go]: pls split this import block

File: java_collection.go [hunk 4]
```diff
@@ -17,10 +17,8 @@
 
 package hessian
 
-import (
-	perrors "github.com/pkg/errors"
-	"reflect"
-)
+import "reflect"
+import perrors "github.com/pkg/errors"
```
Comment (AlexStocks) [inline java_collection.go]: change it in this way,

import (
    "reflect"
)

import (
     perrors "github.com/pkg/errors"
)

**Comment by AlexStocks** [comment]
LGTM

File: java_collection_test.go (hunk 2 above)
Comment (wongoo) [reply java_collection_test.go]: 我到不觉得是例子，一旦加入实际上则成为了标准，用户也不会另外实现。 所以我也建议尝试复用gost里面的hashset。

**Pull request merged**